GetMessages/handler.ts

```diff
@@ -50,42 +65,105 @@ type IGetMessagesHandlerResponse =
-  return async fiscalCode => {
-    return pipe(
-      messageModel.findMessages(fiscalCode),
-      TE.map(flattenAsyncIterator),
-      TE.map(_ => filterAsyncIterator(_, isRight)),
-      TE.map(_ => mapAsyncIterator(_, e => e.right)),
-      TE.map(_ => filterAsyncIterator(_, RetrievedNotPendingMessage.is)),
-      TE.map(_ => mapAsyncIterator(_, retrievedMessageToPublic)),
-      TE.bimap(
-        failure => ResponseErrorQuery(failure.kind, failure),
-        ResponseJsonIterator
-      ),
-      TE.toUnion
-    )();
-  };
-}
+export const GetMessagesHandler = (
+  messageModel: MessageModel,
+  serviceModel: ServiceModel,
+  blobService: BlobService
+): IGetMessagesHandler => async (
+  fiscalCode,
+  maybePageSize,
+  maybeEnrichResultData,
+  maybeMaximumId,
+  maybeMinimumId
+) =>
+  pipe(
+    T.Do,
+    T.bind("pageSize", () =>
+      T.of(O.getOrElse(() => defaultPageSize)(maybePageSize))
+    ),
+    T.bind("shouldEnrichResultData", () =>
+      T.of(O.getOrElse(() => false)(maybeEnrichResultData))
+    ),
+    T.bind("maximumId", () => T.of(O.toUndefined(maybeMaximumId))),
+    T.bind("minimumId", () => T.of(O.toUndefined(maybeMinimumId))),
+    T.map(({ pageSize, shouldEnrichResultData, maximumId, minimumId }) =>
+      pipe(
+        messageModel.findMessages(fiscalCode, pageSize, maximumId, minimumId),
+        TE.map(i => mapAsyncIterator(i, A.rights)),
+        TE.map(i =>
+          mapAsyncIterator(i, A.filter(RetrievedNotPendingMessage.is))
+        ),
+        TE.map(i => mapAsyncIterator(i, A.map(retrievedMessageToPublic))),
+        TE.chain(i =>
+          // check whether we should enrich messages or not
+          pipe(
+            TE.fromPredicate(
+              () => shouldEnrichResultData === true,
+              () =>
+                // if no enrichment is requested we just wrap messages in a TE
+                mapAsyncIterator(
+                  i,
+                  // A.map(e => TE.of<Error, CreatedMessageWithoutContent>(e))
+                  A.map(async e =>
+                    E.right<Error, CreatedMessageWithoutContent>(e)
+                  )
+                )
+            )(i),
+            TE.map(j =>
+              mapAsyncIterator(
+                j,
+                enrichMessagesData(messageModel, serviceModel, blobService)
+              )
+            ),
+            TE.orElse(TE.of)
+          )
+        ),
+        TE.map(flattenAsyncIterator),
+        TE.chain(i =>
+          TE.tryCatch(() => asyncIteratorToPageArray(i, pageSize), E.toError)
+        ),
+        TE.chain(
+          TE.fromPredicate(
+            page => !page.results.some(E.isLeft),
+            () => new Error("Cannot enrich data")
+          )
+        ),
+        TE.map(({ hasMoreResults, results }) =>
+          toPageResults(A.rights([...results]), hasMoreResults)
+        ),
+        TE.bimap(e => ResponseErrorInternal(e.message), ResponseSuccessJson),
+        TE.toUnion
+      )()
+    )
+  )();
```

Comment:
You can safely use `TE.bind` instead (this could resolve ResponseErrorQuery too)
```suggestion
  pipe(
    TE.Do,
    TE.bind("pageSize", () =>
      TE.of(O.getOrElse(() => defaultPageSize)(maybePageSize))
    ),
    TE.bind("shouldEnrichResultData", () =>
      TE.of(O.getOrElse(() => false)(maybeEnrichResultData))
    ),
    TE.bind("maximumId", () => TE.of(O.toUndefined(maybeMaximumId))),
    TE.bind("minimumId", () => TE.of(O.toUndefined(maybeMinimumId))),
    TE.chain(({ pageSize, shouldEnrichResultData, maximumId, minimumId }) =>
      pipe(
        messageModel.findMessages(fiscalCode, pageSize, maximumId, minimumId),
        TE.mapLeft(failure => ResponseErrorQuery(failure.kind, failure)),
        TE.map(i => mapAsyncIterator(i, A.rights)),
        TE.map(i =>
          mapAsyncIterator(i, A.filter(RetrievedNotPendingMessage.is))
        ),
        TE.map(i => mapAsyncIterator(i, A.map(retrievedMessageToPublic))),
        TE.chainW(i =>
          // check whether we should enrich messages or not
          pipe(
            TE.fromPredicate(
              () => shouldEnrichResultData === true,
              () =>
                // if no enrichment is requested we just wrap messages in a TE
                mapAsyncIterator(
                  i,
                  // A.map(e => TE.of<Error, CreatedMessageWithoutContent>(e))
                  A.map(async e =>
                    E.right<Error, CreatedMessageWithoutContent>(e)
                  )
                )
            )(i),
            TE.map(_ =>
              mapAsyncIterator(
                _,
                enrichMessagesData(messageModel, serviceModel, blobService)
              )
            ),
            TE.orElse(TE.of),
            TE.map(flattenAsyncIterator),
            TE.chain(i =>
              TE.tryCatch(() => asyncIteratorToPage(i, pageSize), E.toError)
            ),
            TE.chain(
              TE.fromPredicate(
                page => !page.results.some(E.isLeft),
                () => new Error("Cannot enrich data")
              )
            ),
            TE.map(({ hasMoreResults, results }) =>
              toPageResults(A.rights([...results]), hasMoreResults)
            ),
            TE.mapLeft(e => ResponseErrorInternal(e.message))
          )
        ),

        TE.map(ResponseSuccessJson)
      )
    ),
    TE.toUnion
  )();
```